pinot-query-planner/src/main/java/org/apache/pinot/query/planner/logical/PlanNodeToRelConverter.java

```diff
@@ -0,0 +1,468 @@
+import org.apache.pinot.query.planner.plannode.AggregateNode;
+import org.apache.pinot.query.planner.plannode.ExchangeNode;
+import org.apache.pinot.query.planner.plannode.ExplainedNode;
+import org.apache.pinot.query.planner.plannode.FilterNode;
+import org.apache.pinot.query.planner.plannode.JoinNode;
+import org.apache.pinot.query.planner.plannode.MailboxReceiveNode;
+import org.apache.pinot.query.planner.plannode.MailboxSendNode;
+import org.apache.pinot.query.planner.plannode.PlanNode;
+import org.apache.pinot.query.planner.plannode.PlanNodeVisitor;
+import org.apache.pinot.query.planner.plannode.ProjectNode;
+import org.apache.pinot.query.planner.plannode.SetOpNode;
+import org.apache.pinot.query.planner.plannode.SortNode;
+import org.apache.pinot.query.planner.plannode.TableScanNode;
+import org.apache.pinot.query.planner.plannode.ValueNode;
+import org.apache.pinot.query.planner.plannode.WindowNode;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * Converts a {@link PlanNode} into a {@link RelNode}.
+ *
+ * This class is used to convert serialized plan nodes into RelNodes so they can be used when explain with
+ * implementation is requested. Therefore some nodes may be transformed in a way that loses information that is
+ * required to create an actual executable plan but not necessary in order to describe the plan.
+ */
+public final class PlanNodeToRelConverter {
+  private static final Logger LOGGER = LoggerFactory.getLogger(PlanNodeToRelConverter.class);
+
+  private PlanNodeToRelConverter() {
+  }
+
+  public static RelNode convert(RelBuilder builder, PlanNode planNode) {
+    ConverterVisitor visitor = new ConverterVisitor(builder);
+    planNode.visit(visitor, null);
+
+    return visitor.build();
+  }
+
+  private static class ConverterVisitor implements PlanNodeVisitor<Void, Void> {
+    private final RelBuilder _builder;
+
+    public ConverterVisitor(RelBuilder builder) {
+      _builder = builder;
+    }
+
+    private void visitChildren(PlanNode node) {
+      node.getInputs().forEach(input -> input.visit(this, null));
+    }
+
+    @Override
+    public Void visitAggregate(AggregateNode node, Void context) {
+      visitChildren(node);
+
+      try {
+        int[] groupKeyArr = node.getGroupKeys().stream().mapToInt(Integer::intValue).toArray();
+        RelBuilder.GroupKey groupKey = _builder.groupKey(groupKeyArr);
+
+        List<RelBuilder.AggCall> aggCalls =
+            node.getAggCalls().stream().map(functionCall -> RexExpressionUtils.toAggCall(_builder, functionCall))
+                .collect(Collectors.toList());
+
+        _builder.aggregate(groupKey, aggCalls);
+      } catch (RuntimeException e) {
+        LOGGER.warn("Failed to convert aggregate node: {}", node, e);
+        _builder.push(new PinotExplainedRelNode(_builder.getCluster(), "UnknownAggregate", Collections.emptyMap(),
+            node.getDataSchema(), readAlreadyPushedChildren(node)));
+      }
```

Comment:
Given the builder is quite complex, some subtle conditions may produce this error. Given the class javadoc explicitly that the generated tree should be used in explains because semantics may be lost, I think it is fair to return this synthetic node. In the end what we are trying to give users is something like:

```
Execution Plan
LogicalSort(offset=[0], fetch=[10])
  PinotLogicalSortExchange(distribution=[hash], collation=[[]], isSortOnSender=[false], isSortOnReceiver=[false])
    UnknownAggregate
      LogicalJoin(condition=[=($3, $5)], joinType=[inner])
        ...
```

Instead of an error. Being optimistic, I think this will be more useful than just having a blind error.